bpf/nat64.c

```diff
@@ -119,20 +119,31 @@ int nat64(struct __sk_buff* skb)
 		ip.protocol = IPPROTO_ICMP;
 	ip.tot_len = bpf_htons(bpf_ntohs(ip6->payload_len) + sizeof(struct iphdr));
 	if (bpf_ntohs(ip.tot_len) > 1280) // https://mailarchive.ietf.org/arch/msg/behave/JfxCt1fGT66pEtfXKuEDJ8rdd7w/
-		ip.frag_off = bpf_htons(IP_DF);
+	ip.frag_off = bpf_htons(IP_DF);
 	ip.saddr = bpf_htonl(src_addr);
 	ip.daddr = ip6->daddr.in6_u.u6_addr32[3];
 
 	// Calculate the IPv4 one's complement checksum of the IPv4 header.
 	__wsum sum4 = 0;
 	for (int i = 0; i < sizeof(ip) / sizeof(__u16); ++i) {
-			sum4 += ((__u16*)&ip)[i];
+		sum4 += ((__u16*)&ip)[i];
 	}
 	// Note that sum4 is guaranteed to be non-zero by virtue of ip.version == 4
 	sum4 = (sum4 & 0xFFFF) + (sum4 >> 16);  // collapse u32 into range 1 .. 0x1FFFE
 	sum4 = (sum4 & 0xFFFF) + (sum4 >> 16);  // collapse any potential carry into u16
 	ip.check = (__u16)~sum4;                // sum4 cannot be zero, so this is never 0xFFFF
 
+	__u64 diff = 0;
+
+	// calculate checksum difference before any helpers that modify
+	// packet's data are called, because verifier will invalidate
+	// all packet pointers
+	if (ip.protocol == IPPROTO_TCP) {
+		// cover both src and dst address change, buffer for dst addr field is
+		// right after src addr both for IPv4 and IPv6 header
```

Comment:
just add a bit more of info so people can understand the solution, as per offline conversation 

https://www.rfc-editor.org/rfc/rfc2460.html#section-8.1
http://www.tcpipguide.com/free/t_TCPChecksumCalculationandtheTCPPseudoHeader-2.htm

Reply:
Added a rather lengthy explanation in comment.